shared/src/main/scala/mlscript/ConstraintSolver.scala

```diff
@@ -398,18 +398,30 @@ class ConstraintSolver extends NormalForms { self: Typer =>
       }
       
       var first = true
-      val constraintProvenanceHints = tighestLocatedRHS.map { case (isMainProv, r, p) =>
-        if (isMainProv) {
+      val constraintProvenanceHints = tighestLocatedRHS.map { case (r, p) =>
           val msgHead = if (first) msg"Note: constraint arises " else msg""
           first = false
           msg"${msgHead}from ${p.desc}:" -> p.loco
-        }
-        else msg"in the context of ${p.desc}" -> p.loco
       }
       
+      val originProvList = cctx.flatMap{ subCtx =>
+        subCtx.iterator.flatMap { case (l,r) => 
+          //we only one origin prov for each subCtx
+          val considered = List(l optionIf (_.prov.isOrigin), r optionIf (_.prov.isOrigin)).flatten
+          considered
+        }.nextOption()
+      }
+
+      first = true
+      val originProvHints = originProvList.map { l => 
+        val msgHead = if (first) msg"Note: " else msg""
+          first = false
+          msg"${msgHead}Type ${l.prov.desc} is defined at: " -> l.prov.loco 
```

Comment:
This results in messages like "`Type class A is defined at ...`". But a "type class" is something else. Perhaps use this instead:
```suggestion
          msg"${msgHead}${l.prov.desc.capitalize} is defined at: " -> l.prov.loco 
```

Reply:
Yes, it makes sense